### PR TITLE
JClouds 1281- DLO  Swift Implementation

### PR DESCRIPTION
**Pull request by archupsg03**
Have implemented Swift Dynamic Large Object upload strategy. Requires changes in 
1. Manifest upload with X-Object-Manifest header
2. Remove blobs, in order to remove the manifest and segments

### CODE REVIEW

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 1]
```diff
@@ -188,11 +197,17 @@ public boolean createContainerInLocation(Location location, String container) {
...
-         return api.getContainerApi(regionId).create(container, ANYBODY_READ);
+         resultContainerCreate = api.getContainerApi(regionId).create(container, ANYBODY_READ);
       }
-      return api.getContainerApi(regionId).create(container, BASIC_CONTAINER);
+      resultContainerCreate = api.getContainerApi(regionId).create(container, BASIC_CONTAINER);
+      if (resultContainerCreate) {
+         Container val = api.getContainerApi(regionId).get(container);
+         containerCache.put(container, Optional.fromNullable(val));
+      }
+      return resultContainerCreate;
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: How is creating a container related to dynamic large objects?  If this is a separate change, submit a separate pull request.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/SwiftApi.java [hunk 2]
```diff
@@ -17,6 +17,7 @@
 package org.jclouds.openstack.swift.v1;
 
 import java.io.Closeable;
+
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/SwiftApi.java]: Please sweep through the entire pull request and remove these whitespace changes.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 3]
```diff
@@ -387,13 +402,26 @@ public Blob getBlob(String container, String name, GetOptions options) {
...
-      // Multipart objects have a manifest which points to subobjects.  Normally
+      // Multipart objects have a manifest which points to subobjects. Normally
       // deleting a object only deletes the manifest, leaving the subobjects.
       // We first try a multipart delete and if that fails since the object is
       // not an MPU we fall back to single-part delete.
-      DeleteStaticLargeObjectResponse response = api.getStaticLargeObjectApi(regionId, container).delete(name);
-      if (!response.status().equals("200 OK")) {
+
+      String objManifest = getManifestInfo(container, name);
+      String sloInfo = getSLOInfo(container, name);
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Previously this code issued one or two DELETEs depending if the name was an SLO or a regular object.  With this commit we issue always issue three RPCs, getWithoutBody, getWithoutBody, and delete with some flags.  How can you optimize this?  Also this fails if the object does not exist.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 4]
```diff
@@ -519,10 +557,19 @@ public String completeMultipartUpload(MultipartUpload mpu, List<MultipartPart> p
          String path = mpu.containerName() + "/" + getMPUPartName(mpu, part.partNumber());
          builder.add(Segment.builder().path(path).etag(part.partETag()).sizeBytes(part.partSize()).build());
       }
-
       return api.getStaticLargeObjectApi(regionId, mpu.containerName()).replaceManifest(mpu.blobName(),
             builder.build(), mpu.blobMetadata().getUserMetadata(), getContentMetadataForManifest(mpu.blobMetadata().getContentMetadata()));
    }
+   
+   public String completeMultipartUploadDLO(MultipartUpload mpu, List<MultipartPart> parts) {
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Remove - we should not expose dynamic large objects in the portable abstraction!  Users can call the provider-level APIs for this functionality.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 5]
```diff
@@ -908,4 +957,115 @@ public void run() {
...
+    * @param container
+    * @param objectName
+    * @return true if object is SLO uploaded else null
+    */
+   private String getSLOInfo(String container, String objectName) {
+      ObjectApi objectApi = api.getObjectApi(regionId, container);
+      Multimap<String, String> objInfo = objectApi.getWithoutBody(objectName).getHeaders();
+      if (objInfo.containsKey("X-Static-Large-Object")) {
+         Collection<String> objectManifest = objInfo.get("X-Static-Large-Object");
+         return String.valueOf(objectManifest.toArray()[0]);
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Consider the simpler:

```java
ObjectApi objectApi = api.getObjectApi(regionId, container);
Multimap<String, String> objInfo = objectApi.getWithoutBody(objectName).getHeaders();
Collection<String> objectManifest = objInfo.get("X-Static-Large-Object");
return Iterables.getFirst(objectManifest, null);
```

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 6]
```diff
@@ -908,4 +957,115 @@ public void run() {
...
+      }
+      return null;
+   }
+   
+   
+   /**
+    * @param containerAndPrefix
+    */
+   private void removeObjectsWithPrefix(String containerAndPrefix) {
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Add `static`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 7]
```diff
@@ -908,4 +957,115 @@ public void run() {
...
+      return parts;
+   }
+    
+   /**
+    * Retrieves a list of existing storage
+    * @param container
+    * @param options
+    * @return maximum of 10,000 object names per request.
+    */
+   private PageSet<? extends StorageMetadata> list(final String container,
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Why do you reimplement `ObjectApi.list`?

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.openstack.swift.v1.options;
+
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.jclouds.blobstore.options.PutOptions;
+
+public class OpenStackSwiftPutOptions extends PutOptions {
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: Remove this entire thing and make users call either the dynamic or static APIs.

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+   public void setup() {
+      super.setup();
+      for (String regionId : regions) {
+         boolean created = getApi().getContainerApi(regionId).create(defaultContainerName);
+         if (!created) {
+            deleteAllObjectsInContainer(regionId, defaultContainerName);
+         }
+      }
+
+      megOf1s = new byte[1024 * 1024];
```
Comment (andrewgaul) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java]: Call `TestUtils.randomByteSource()` instead.

File: blobstore/src/main/java/org/jclouds/blobstore/options/PutOptions.java [hunk 10]
```diff
@@ -39,11 +39,11 @@
...
 
    private BlobAccess blobAccess = BlobAccess.PRIVATE;
-   private boolean multipart = false;
-   private boolean useCustomExecutor = false;
+   protected boolean multipart = false;
+   protected boolean useCustomExecutor = false;
 
    // TODO: This exposes ListeningExecutorService to the user, instead of a regular ExecutorService
-   private ListeningExecutorService customExecutor = MoreExecutors.sameThreadExecutor();
+   protected ListeningExecutorService customExecutor = MoreExecutors.sameThreadExecutor();
```
Comment (andrewgaul) [inline blobstore/src/main/java/org/jclouds/blobstore/options/PutOptions.java]: Revert this after removing `OpenStackSwiftPutOptions`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 11]
```diff
@@ -230,10 +245,10 @@ public void setContainerAccess(String name, ContainerAccess access) {
       ObjectApi objectApi = api.getObjectApi(regionId, container);
       ObjectList objects = objectApi.list(toListContainerOptions.apply(options));
       if (objects == null) {
-         containerCache.put(container, Optional.<Container> absent());
+         //containerCache.put(container, Optional.<Container> absent());
          return new PageSetImpl<StorageMetadata>(ImmutableList.<StorageMetadata> of(), null);
       } else {
-         containerCache.put(container, Optional.of(objects.getContainer()));
+         //containerCache.put(container, Optional.of(objects.getContainer()));
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Why comment this out?

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/features/DynamicLargeObjectApi.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    *           name/key.
+    *
+    * @return {@link SwiftObject#getEtag()} of the object, which is the MD5
+    *         checksum of the concatenated ETag values of the {@code segments}.
+    */
+   @Named("dynamicLargeObject:replaceManifest")
+   @PUT
+   @ResponseParser(ETagHeader.class)
+   @Headers(keys = "X-Object-Manifest", values = "{containerName}/{objectName}/")
+   String replaceManifest(@PathParam("containerName") String containerName, @PathParam("objectName") String objectName,
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/features/DynamicLargeObjectApi.java]: Annotate this method @Deprecated and document that users should prefer SLO.

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStoreParallelLiveTest.java [hunk 13]
```diff
@@ -123,6 +124,50 @@ public void uploadMultipartBlob() {
...
+            .payload(new FilePayload(bigFile))
+            .build();
+      // configure the blobstore to use multipart uploading of the file
+      OpenStackSwiftPutOptions opt = new OpenStackSwiftPutOptions();
+      opt.multipart(executor);
+      opt.DLO(true);
+      String eTag = blobStore.putBlob(CONTAINER, blob, opt);
+      // assertEquals(eTag, etag);
+      // The etag returned by Swift is not the md5 of the Blob uploaded
+      // It is the md5 of the concatenated segment md5s
```
Comment (andrewgaul) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStoreParallelLiveTest.java]: Why not verify the ETag?

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+      String etag1s = objectApi.put(name + "/1", newByteSourcePayload(ByteSource.wrap(megOf1s)));
+      awaitConsistency();
+      assertMegabyteAndETagMatches(regionId, containerName, name + "/1", etag1s);
+
+      String etag2s = objectApi.put(name + "/2", newByteSourcePayload(ByteSource.wrap(megOf2s)));
+      awaitConsistency();
+      assertMegabyteAndETagMatches(regionId, containerName, name + "/2", etag2s);
+
+      List<Segment> segments = ImmutableList.<Segment> builder()
```
Comment (andrewgaul) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java]: Prefer `ImmutableList.of`.

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+      ObjectList objects = getApi().getObjectApi(regionId, containerName).list(new ListContainerOptions());
+      if (objects == null) {
+         return;
+      }
+      List<String> pathsToDelete = Lists.transform(objects, new Function<SwiftObject, String>() {
+         public String apply(SwiftObject input) {
+            return containerName + "/" + input.getName();
+         }
+      });
+      if (!pathsToDelete.isEmpty()) {
```
Comment (andrewgaul) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java]: Unneeded conditional.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 6 above)
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Remove this function and make the single caller explicitly call `splitContainerAndKey` and `removeObjectsWithPrefix(String, String)`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 16]
```diff
@@ -908,4 +957,115 @@ public void run() {
...
+      do {
+         org.jclouds.openstack.swift.v1.options.ListContainerOptions listContainerOptions = prefix(prefix);
+         if (nextMarker != null) {
+            listContainerOptions = listContainerOptions.marker(nextMarker);
+         }
+         PageSet<? extends StorageMetadata> chunks = list(container, listContainerOptions);
+         Iterator<? extends StorageMetadata> iter = chunks.iterator();
+
+         while (iter.hasNext()) {
+            objectApi.delete(iter.next().getName());
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Call `BulkApi.buckDelete`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java [hunk 17]
```diff
@@ -908,4 +957,115 @@ public void run() {
...
+
+      removeObjectsWithPrefix(container, prefix);
+   }
+
+    
+   /**
+    * @param containerAndKey
+    * @return array of size 2 with container and objectPrefix
+    */
+   private String[] splitContainerAndKey(String containerAndKey) {
```
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Add `static`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 7 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Deleting DLO objects requires, the details of the segments.

BlobStore's List command takes BlobStore's ListContainerOptions as input.
But incase of remove blob we need Swift's ListContainerOptions. That the reasonf i copied and made few changes

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStoreParallelLiveTest.java (hunk 13 above)
Comment (archupsg03) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStoreParallelLiveTest.java]: Etag comparison was already commented in the previous function

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java (hunk 14 above)
Comment (archupsg03) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java]: These are just copied from SLO testcase, and appropriate changes has been made for DLO. These things remain the same.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 1 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: This is a separate fix. Will submit another PR

**Comment by archupsg03** [comment]
For DLO upload, etag will the md5 of the manifest and not the content of the file thats getting uploaded.
Thats the reason etag mismatch happens in ParallelLiveTest

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java (hunk 8 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: Am not getting this point :(

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 4 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Not getting this point too

Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: jclouds consists of two levels, the lower-level provider APIs, e.g., `DynamicLargeObjectApi`, and the higher-level portable abstraction, e.g., `RegionScopedSwiftBlobStore`.  The former exposes all the functionality of a given provider, e.g, SLO, DLO, etc., while the latter only implements a portable layer which is common between providers.  Thus `completeMultipartUploadDLO` should not exist and if callers want this functionality they can use `RegionScopedSwiftBlobStore`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 7 above)
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: `RegionScopedSwiftBlobStore` can call `ObjectApi.list`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java (hunk 8 above)
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: Once you address https://github.com/jclouds/jclouds/pull/1090#discussion_r114451055 this code becomes unnecessary.

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java (hunk 14 above)
Comment (andrewgaul) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java]: If the inconsistency bothers you, please submit a fix for the existing SLO tests as well.

**Comment by andrewgaul** [comment]
> For DLO upload, etag will the md5 of the manifest and not the md5 of the concatenated etags.
> Thats the reason etag mismatch happens in ParallelLiveTest. So etag checking for DLO objects requires
> manifest to be constructed locally and compute its md5 and then compare with the etag from Swift

We cannot introduce failing tests so how do you plan to address this?

File: apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.io.ByteSource;
+import com.google.common.util.concurrent.Uninterruptibles;
+
+@Test(groups = "live", testName = "DynamicLargeObjectApiLiveTest", singleThreaded = true)
+public class DynamiclargeObjectApiLiveTest extends BaseSwiftApiLiveTest {
```
Comment (andrewgaul) [inline apis/openstack-swift/src/test/java/org/jclouds/openstack/swift/v1/features/DynamiclargeObjectApiLiveTest.java]: Rename class to `DynamicLargeObjectApiLiveTest` with a capital "L".

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 7 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Theres problem in using objectApi.List. I am not able to get the markers. 

How does objecApi.List handles once it lists the first 10k objects.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java (hunk 8 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: Say if the user prefers to call putMultipart Blob with DLO, doesn't want to use DLO API. How can such cases be handled ?

Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: The portable abstraction does not expose all functionality.  The caller must use the DLO API to access DLO.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 7 above)
Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: `ObjectApi.list(ListContainerOptions)` Javadoc says:

> Lists up to 10,000 objects. To control a large list of containers beyond 10,000 objects, use the `marker` and `endMarker` parameters in the `ListContainerOptions` class.

**Comment by andrewgaul** [comment]
@archupsg03 The more I think about this, let's work on the portable abstraction code in a separate pull request.  Please focus this on only the provider, i.e., remove `RegionScopedSwiftBlobStore` changes and only add `RegionScopedSwiftBlobStore`.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java (hunk 16 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Does DLO, bulk deletes  are also expected to delete only the manifest file ?

Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java]: Yes.  DLO generally is poorly thought and and supported.  Again, scope this commit down to just the provider changes and submit the portable abstraction changes as a separate pull request.

File: apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java (hunk 8 above)
Comment (archupsg03) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: I am almost done with rest of the changes, except the separation of lower and higher API.
My understanding from the comment #1090, is that DynamicLargeObjetApi should have methods such as initiateMultiPartUpload, uploadMultiPart, completeMultiPart etc., Is this correct ? If not
 Can you just give an example, say what methods are expected to in DynamiCLargeObjectApi file  ?

Comment (andrewgaul) [inline apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/options/OpenStackSwiftPutOptions.java]: No, you still don't understand the difference between the provider API and the portable abstraction.  Go compare something like B2 with its [MultipartApi](https://github.com/jclouds/jclouds/blob/master/providers/b2/src/main/java/org/jclouds/b2/features/MultipartApi.java) against [B2BlobStore](https://github.com/jclouds/jclouds/blob/master/providers/b2/src/main/java/org/jclouds/b2/blobstore/B2BlobStore.java).  The former implements the REST API while the latter conforms to the `BlobStore` interface.  This pull request should not contain any of the latter.

**Comment by archupsg03** [comment]
No changes are required in RegionScopedBlobStore. Working on the DynamicLargeObjectAPI. Will submit PR for that. Closing this PR.
